### PR TITLE
Expenses transaction analyzer

### PR DESCRIPTION
**Pull request by Azher9767**

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -7,3 +7,69 @@
...
+category = Category.create(name: "utility")
+Category.create(name: "electricity", parent_id: category.id)
+Category.create(name: "gas", parent_id: category.id)
+Category.create(name: "maintenance", parent_id: category.id)
+Category.create(name: "mobile", parent_id: category.id)
+Category.create(name: "insurance", parent_id: category.id)
+Category.create(name: "others", parent_id: category.id)
+
+category = Category.create(name: "petrol")
+Category.create(name: "petrol", parent_id: category.id)
```
Comment (ahmad19) [inline db/seeds.rb:51]: Remove this.

File: app/controllers/transactions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    @transactions = Transaction.all
+  end
+
+  def show 
+    @transaction = Transaction.find(params[:id])
+    @string_to_hash = JSON.parse(@transaction.data)
+
+    obj = CategoryAnalyzer.new(@string_to_hash)
+    @per_category = obj.expenses_per_category
+    @per_subcategory = obj.expenses_per_subcategory
```
Comment (ahmad19) [inline app/controllers/transactions_controller.rb:12]: Move this code block in the transaction helper as it is only being used in the views.

File: app/controllers/transactions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  def new
+    @transaction = Transaction.new
+  end
+
+  def create
+    require 'csv'
+    @result = []
+    CSV.foreach((params[:csv_file]), headers: true, col_sep: ",") do |row|
+      @result << row
+    end
```
Comment (ahmad19) [inline app/controllers/transactions_controller.rb:24]: Move this code block into another class, may be ?

File: app/helpers/transaction_helper.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    total = categories_amount.values.sum.to_f
+    categories_amount.transform_values { |value| ((value/total) * 100).round(1) }
+  end
+
+  def per_subcategory_percentage(categories_with_children)
+    categories_with_children.each_with_object({}) do |(category, subcategories_with_amounts), result|
+        result[category] = subcategories_with_amounts.each_with_object({}) do |(sub_category, amount), subcategory_result|
+           total_amount_category = subcategories_with_amounts.values.sum.to_f
+           subcategory_result[sub_category] = ((amount /  total_amount_category) * 100).round(2)
+        end
```
Comment (ahmad19) [inline app/helpers/transaction_helper.rb:12]: Correct the alignment

File: db/seeds.rb [hunk 5]
```diff
@@ -7,3 +7,69 @@
...
+
+category = Category.create(name: "grocery")
+Category.create(name: "fruit", parent_id: category.id)
+Category.create(name: "dudaram", parent_id: category.id)
+Category.create(name: "others", parent_id: category.id)
+
+category = Category.create(name: "personal")
+Category.create(name: "bike", parent_id: category.id)
+Category.create(name: "insurance", parent_id: category.id)
+Category.create(name: "others", parent_id: category.id)
```
Comment (ahmad19) [inline db/seeds.rb:61]: Add petrol in personal.

File: db/seeds.rb [hunk 6]
```diff
@@ -7,3 +7,69 @@
...
+Category.create(name: "salary", parent_id: category.id)
+Category.create(name: "coworking", parent_id: category.id)
+Category.create(name: "others", parent_id: category.id)
+
+category = Category.create(name: "others")
+
+
+
+
+
```
Comment (ahmad19) [inline db/seeds.rb:75]: No need of these extra lines

File: spec/fixtures/expenses_data.csv [hunk 7]
```diff
@@ -0,0 +1,9 @@
+Date,Narration,Chq./Ref.No.,Value Dt,Withdrawal Amt.,Deposit Amt.
```
Comment (ahmad19) [inline spec/fixtures/expenses_data.csv:1]: remove this file.

**Review by ahmad19 — CHANGES_REQUESTED** [review]
Make changes

File: db/seeds.rb (hunk 1 above)
Comment (Azher9767) [reply db/seeds.rb:51]: done

File: app/views/transactions/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                  <% end %>
+                </td>
+              </tr>
+            <% end %>  
+          </tbody>
+        </table>
+      </div>
+    </div>
+  </div>  
+<% end %>
```
[the remark this answers is not in the document]
Comment (Azher9767) [reply app/views/transactions/show.html.erb:65]: done

File: db/seeds.rb (hunk 5 above)
Comment (Azher9767) [reply db/seeds.rb:61]: done

File: spec/fixtures/expenses_data.csv (hunk 7 above)
Comment (Azher9767) [reply spec/fixtures/expenses_data.csv:1]: done

File: app/controllers/transactions_controller.rb (hunk 2 above)
Comment (Azher9767) [reply app/controllers/transactions_controller.rb:12]: done

File: app/helpers/transaction_helper.rb [hunk 9]
```diff
@@ -13,13 +13,15 @@ def per_subcategory_percentage(categories_with_children)
     end
   end
 
+  def string_to_hash  
+    @string_to_hash = JSON.parse(@transaction.data)
+  end
+
```
Comment (ahmad19) [inline app/helpers/transaction_helper.rb:19]: This needs to be memoized.

**Review by ahmad19 — APPROVED** [review]

**Pull request merged**